src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs

```diff
@@ -1201,27 +1201,26 @@ public void Dispose()
             {
             }
 
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
             public bool MoveNext()
             {
                 List<T> localList = _list;
 
-                if (_version == localList._version && ((uint)_index < (uint)localList._size))
+                if (_version != localList._version)
                 {
-                    _current = localList._items[_index];
-                    _index++;
-                    return true;
+                    ThrowHelper.ThrowInvalidOperationException_InvalidOperation_EnumFailedVersion();
                 }
-                return MoveNextRare();
-            }
 
-            private bool MoveNextRare()
-            {
-                if (_version != _list._version)
+                T[] items = localList._items;
+                int index = _index;
+                if ((uint)index < (uint)items.Length &&
+                    (uint)index < (uint)localList._size)
```

Comment:
Maybe reordering these compares could make it slightly faster since the 2nd comparison is more restrictive. We could also do an unsafe array access to avoid the 1st comparison/bounds check altogether...

Reply:
> I wonder if the JIT could recognize such comparisons and fold them together (and whether that would help)? If I were to rewrite it as `(uint)index < Math.Min((uint)items.Length, (uint)localList._size)`, the min gets emitted via branchless instruction, but the JIT then loses the connection with the indexer and emits the bounds check.

can you file an issue with a repro please?

Reply:
> What happens if we remove the check of (uint)index < (uint)items.Length?

Then the same check and branch happens implicitly as part of the bounds check.

Reply:
And then? Potential performance loss? I would prefer removing the manual check if it doesn't affect performance, because it can bring a better behavior that throwing an exception.